### PR TITLE
Optimizing inducing points

### PR DESCRIPTION
**Pull request by yorkerlin**
@karlnapf 
take a look at this

### CODE REVIEW

**Comment by yorkerlin** [comment]
@karlnapf 
may take a look at the notebook for demonstration

http://nbviewer.ipython.org/gist/yorkerlin/e69926dbea6d47946d79#optInd

File: src/shogun/machine/gp/SingleFITCLaplacianBase.cpp [hunk 1]
```diff
@@ -404,4 +429,124 @@ SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inducing_featur
...
+	check_fully_FITC();
+	REQUIRE(m_fully_FITC,"Please use a kernel which supports to optimize inducing features\n");
+
+	//features by samples
+	SGMatrix<float64_t>& lat_m=m_inducing_features;
+	SGVector<double> x(lat_m.matrix,lat_m.num_rows*lat_m.num_cols,false);
+
+	// create nlopt object and choose MMA (Method of Moving Asymptotes)
+	// optimization algorithm
+	nlopt_opt opt=nlopt_create(NLOPT_LD_LBFGS, lat_m.num_rows*lat_m.num_cols);
```
Comment (yorkerlin) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.cpp]: In fact, we can choose difference optimization algorithms.

File: src/shogun/machine/gp/SingleFITCLaplacianBase.cpp [hunk 2]
```diff
@@ -404,4 +429,124 @@ SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inducing_featur
...
+	check_fully_FITC();
+	REQUIRE(m_fully_FITC,"Please use a kernel which supports to optimize inducing features\n");
+
+	//features by samples
+	SGMatrix<float64_t>& lat_m=m_inducing_features;
+	SGVector<double> x(lat_m.matrix,lat_m.num_rows*lat_m.num_cols,false);
+
+	// create nlopt object and choose LBFGS
+	// optimization algorithm
+	nlopt_opt opt=nlopt_create(NLOPT_LD_LBFGS, lat_m.num_rows*lat_m.num_cols);
```
Comment (yorkerlin) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.cpp]: In fact, we can choose difference optimization algorithms.

File: src/shogun/machine/gp/FITCInferenceBase.cpp [hunk 3]
```diff
@@ -50,45 +49,43 @@ CFITCInferenceBase::CFITCInferenceBase() : CInferenceMethod()
...
 }
 
 void CFITCInferenceBase::convert_features()
 {
 	CDotFeatures *feat_type=dynamic_cast<CDotFeatures *>(m_features);
-	CDotFeatures *lat_type=dynamic_cast<CDotFeatures *>(m_inducing_features);
+
+	SGMatrix<float64_t>lat_m(m_inducing_features.matrix,
+		m_inducing_features.num_rows,m_inducing_features.num_cols,false);
+	CDotFeatures *lat_type=new CDenseFeatures<float64_t>(lat_m);
```
Comment (karlnapf) [inline src/shogun/machine/gp/FITCInferenceBase.cpp]: maybe lat_type should be changed to inducing_feat_type at some point

File: src/shogun/machine/gp/FITCInferenceBase.cpp [hunk 4]
```diff
@@ -109,7 +106,7 @@ void CFITCInferenceBase::init()
 	SG_ADD(&m_Sigma, "Sigma", "covariance matrix of the approximation to the posterior", MS_NOT_AVAILABLE);
 	SG_ADD(&m_ktrtr_diag, "ktrtr_diag", "diagonal elements of kernel matrix m_ktrtr", MS_NOT_AVAILABLE);
 
-	m_inducing_features=NULL;
+	m_inducing_features=SGMatrix<float64_t>();
```
Comment (karlnapf) [inline src/shogun/machine/gp/FITCInferenceBase.cpp]: Do we start like this?
Not rather a random subset of the training data? Or is this done later?

File: src/shogun/machine/gp/FITCInferenceBase.cpp [hunk 5]
```diff
@@ -126,16 +123,14 @@ float64_t CFITCInferenceBase::get_inducing_noise()
...
 }
 
 void CFITCInferenceBase::check_members() const
 {
 	CInferenceMethod::check_members();
 
-	REQUIRE(m_inducing_features, "Inducing features should not be NULL\n")
-	REQUIRE(m_inducing_features->get_num_vectors(),
-			"Number of inducing features must be greater than zero\n")
+	REQUIRE(m_inducing_features.num_rows, "Inducing features should not be empty\n")
```
Comment (karlnapf) [inline src/shogun/machine/gp/FITCInferenceBase.cpp]: please add row/col to the error message

File: src/shogun/machine/gp/SingleFITCLaplacianBase.cpp [hunk 6]
```diff
@@ -63,7 +68,27 @@ void CSingleFITCLaplacianBase::init()
...
+	SG_ADD(&m_lower_bound, "lower_bound",
+		"lower bound of inducing features", MS_NOT_AVAILABLE);
+	SG_ADD(&m_max_ind_iterations, "max_ind_iterations",
+		"max number of iterations used in inducing features optimization", MS_NOT_AVAILABLE);
+	SG_ADD(&m_ind_tolerance, "ind_tolerance",
+		"tolearance used in inducing features optimization", MS_NOT_AVAILABLE);
+	SG_ADD(&m_opt_inducing_features,
+		"opt_inducing_features", "whether optimize inducing features", MS_NOT_AVAILABLE);
+
+	m_max_ind_iterations=50;
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.cpp]: Those parameters should be documented in the class docs

File: src/shogun/machine/gp/SingleFITCLaplacianBase.cpp [hunk 7]
```diff
@@ -404,4 +432,124 @@ SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inducing_featur
...
+	m_opt_inducing_features=is_optmization;
+}
+
+void CSingleFITCLaplacianBase::optimize_inducing_features()
+{
+	if (!m_opt_inducing_features)
+		return;
+
+	check_fully_FITC();
+	REQUIRE(m_fully_FITC,"Please use a kernel which supports to optimize inducing features\n");
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.cpp]: Maybe print name of kernel?

File: src/shogun/machine/gp/SingleFITCLaplacianBase.cpp [hunk 8]
```diff
@@ -404,4 +432,124 @@ SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inducing_featur
...
+
+	if (m_lower_bound.vlen>0)
+	{
+		SGVector<double> lower_bound(lat_m.num_rows*lat_m.num_cols);
+		if(m_lower_bound.vlen==1)
+			lower_bound.set_const(m_lower_bound[0]);
+		else
+		{
+			for(index_t j=0; j<lat_m.num_cols; j++)
+				std::copy(m_lower_bound.vector, m_lower_bound.vector+m_lower_bound.vlen,
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.cpp]: can't this be done in a single memcpy call?

File: src/shogun/machine/gp/SingleFITCLaplacianBase.h [hunk 9]
```diff
@@ -99,7 +99,80 @@ class CSingleFITCLaplacianBase: public CFITCInferenceBase
 	 * @param kern kernel to set
 	 */
 	virtual void set_kernel(CKernel* kern);
+
+#ifdef HAVE_NLOPT
+	/** opitmize inducing features
+	 */
+	virtual void optimize_inducing_features();
+
+	/** set the lower bound of inducing features
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.h]: Set with capital S

File: src/shogun/machine/gp/SingleFITCLaplacianBase.h [hunk 10]
```diff
@@ -99,7 +99,80 @@ class CSingleFITCLaplacianBase: public CFITCInferenceBase
...
+	 *
+	 * @param it max number of iterations
+	 */
+	virtual void set_max_iterations_for_inducing_features(int32_t it);
+
+	/** whether enable to opitmize inducing features
+	 *
+	 * @param is_optmization enable optimization
+	 */
+	virtual void enable_optimizing_inducing_features(bool is_optmization);
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleFITCLaplacianBase.h]: set_optimize_inducing_features(bool optimize_inducing_features)

**Comment by karlnapf** [comment]
Fine to merge this one apart from some technical minors

**Comment by karlnapf** [comment]
Ill merge since travis error is unrelated.
BTW are there any unit tests for this?

**Comment by karlnapf** [comment]
As usual, it would be cool to have a patch that addresses the comments above
Just trying it on my machine

**Pull request merged**